### ISSUE TITLE
Autotools has stopped working

### ISSUE THREAD

**Issue by rsgabriel** [Novik/ruTorrent#2139]
For about 6 months rTorrent and Autotools' "path to finished downloads" and "add label to path" functionality has been running  smoothly for me.

From time to time it would seem like it would stop working, and a restart would fix it, but now I'm at a point where a restart does nothing and I have to manually move downloads.

I switched from move to hard link which also fails.

I enabled every log channel I could find online and into ~/.rtorrent.rc, including debug, but it seems there is no information in the log emitted by Autotools.

How can I troubleshoot this?

**Comment by rsgabriel** [comment]
I found the Autotools log, and enabled debug output:

Not sure what to make of this:

> [31.01.21 23:43:50] 
> [31.01.21 23:43:50] AutoMove: --- begin ---
> [31.01.21 23:43:50] AutoMove: --- end ---
> [31.01.21 23:45:06] <?xml version="1.0" encoding="UTF-8"?><methodCall><methodName>system.multicall</methodName><params><param><value><array><data>
> <value><struct><member><name>methodName</name><value><string>get_up_total</string></value></member><member><name>params</name><value><array><data>
> </data></array></value></member></struct></value>
> <value><struct><member><name>methodName</name><value><string>get_down_total</string></value></member><member><name>params</name><value><array><data>
> </data></array></value></member></struct></value>
> <value><struct><member><name>methodName</name><value><string>d.multicall</string></value></member><member><name>params</name><value><array><data>
> <value><string>main</string></value>
> <value><string>d.get_hash=</string></value>
> <value><string>d.get_up_total=</string></value>
> <value><string>d.get_down_total=</string></value>
> </data></array></value></member></struct></value>
> </data></array></value></param></params></methodCall>
> [31.01.21 23:45:06] Status: 200 OK
> Content-Type: text/xml
> Content-Length: 812
> 
> <?xml version="1.0" encoding="UTF-8"?>
> <methodResponse>
> <params>
> <param><value><array><data>
> <value><struct>
> <member><name>faultCode</name>
> <value><i4>-506</i4></value></member>
> <member><name>faultString</name>
> <value><string>Method 'get_up_total' not defined</string></value></member>
> </struct></value>
> <value><struct>
> <member><name>faultCode</name>
> <value><i4>-506</i4></value></member>
> <member><name>faultString</name>
> <value><string>Method 'get_down_total' not defined</string></value></member>
> </struct></value>
> <value><struct>
> <member><name>faultCode</name>
> <value><i4>-506</i4></value></member>
> <member><name>faultString</name>
> <value><string>Method 'd.multicall' not defined</string></value></member>
> </struct></value>
> </data></array></value></param>
> </params>
> </methodResponse>
>

**Comment by Novik** [comment]
In your case ruTorrent can't determine version of rtorrent correctly.
Reason of this may be vary.
For example, you have modified version of rtorrent. Or incorrect permission of 'share' ruTorrent directory. Or something like this.

**Comment by rsgabriel** [comment]
Hi Novik,

>For example, you have modified version of rtorrent.

I use(d) the rtinst script for installation:

https://github.com/arakasi72/rtinst

> Or incorrect permission of 'share' ruTorrent directory. 

To be clear, this has been working for months with no need to login to the seedbox until it stops working. Also, I have temporarily resolved the issue by rebooting the machine with no permissions changes and am now seeing both Hard Link and Move options working again.

>Or something like this.

For radarr/sonarr sometimes rTorrent becomes unresponsive. I am seeding 41TB across 5500 files. Could this be a situation where when autotools runs, it is timing out, and then never runs again for the same torrent? Is some kind of retry mechanism available?


It would be nice to avoid this issue in the future, I have added some additional logging to assist in determine the flow of what is happening, but any ideas/tips you might have would be appreciated.

**Comment by rsgabriel** [comment]
I have some results from the extended logging I setup.

I found 34 files in both sonarr/radarr that had not moved with many others that had.

I selected a torrent that had failed to moved and inspected the log file:

https://pastebin.com/qn7Jus4R

Notably:

> AutoMove: -session=
> AutoMove: path_to_finished if block was NOT entered

Searching for the string "-session" in the log file confirms that all 34 of these incidents occurred sequentially, this is observable by these search results:

https://pastebin.com/VaVePLT3

The issue began [11.02.21 19:56:54] and the final non-working one was [13.02.21 02:23:39] following by two working moves the first of which was 20m after the final failure. In other words, the issue comes and goes.

So the failure is because:

$session = rTorrentSettings::get()->session;

...returns an empty string.

**Comment by rsgabriel** [comment]
So the questions become:

1) Can this be prevented?

2) If not, how can it be mitigated? Can a queue be created and items are only removed upon success? Can the rpc request that gets the session retry until it is successful?

**Comment by Novik** [comment]
If you have properly configured rtorrent/ruTorrent, then when rtorrent is started (and\or WebUI is started) the ruTorrent make initialization and store all properties in the file ```share/users/USERNAME/settings/rtorrent.dat```. Include path to session, version of rtorrent etc.
I think in your case this file is absent/empty/contains crap. I have no idea - why.
This may happens, for example, if at [11.02.21 19:56:54] someone delete this file and at [13.02.21 02:42:23] rtorrent was restarted (or WebUI was loaded).

**Comment by rsgabriel** [comment]
> I think in your case this file is absent/empty/contains crap

upon checking, the file appears to contain valid json

> someone delete this file

No one did this, I am the only person with access to both the physical and virtual machine rTorrent runs on.

> rtorrent was restarted 

rTorrent is not configured to automatically restart, so this did not happen.

>(or WebUI was loaded).

This did occur.

As a side note, I had not logged in to ruTorrent since my last comment and everything has moved correctly since then

Any thoughts about what this could be otherwise? Should I just hard code the path in?

**Comment by rsgabriel** [comment]
After resolving the issue in #2162 things have been running pretty smoothly again, however I discovered that for the last week no automoves have been occurring.

After logging in to the web ui I saw in my log file that a single move occurred, however I have 116 torrents that have been left unmoved.

I assume that by opening the web ui that I kicked off autotools somehow which resulted in the move of the single torrent. 

The question then becomes:

How can I automate the launching of the autotools or detect it has crashed / exited?
